laurat007/assignment3/Word.java

```diff
@@ -0,0 +1,243 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.TreeSet;
+
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+/**
+ *
+ * @author Laura
+ * 
+ *  This program returns the words that can be formed using the letters in a 
+ * given grid by moving left, right, up, down and on diagonals. 
+ *  I used a dictionary class in order to define the correspondence between a 
+ * particular word and its explanation ( + the 2 helping methods isPrefix and 
+ * isWord, and a grid class in order to define a cell (a pair of indexes - which 
+ * facilitates moving in the grid) and its value.
+ *  The Cell subclass has methods implemented for every moving option.
+ *  Moving in the grid is similar to a backtracking from each Cell which stops
+ * if the current word formed is not a prefix, and if it finds a word it is added
+ * to solution.
+ * 
+ */
+
+class Dictionary {
```

Comment:
This Dictionary could be improved a lot. As I mentioned yesterday in our meeting: even though it is not a requirement I strongly encourage you to try and come up with a more efficient solution. A hint: lookup time for isWord or isPrefix should have the maximum complexity of the length of the word or prefix that is looked up (or the longest word in the dictionary, whichever is shorter). In your case the current complexity s the number of words in the dictionary, which can potentially be huge.
Let me know if you decide to give it a try and need a hint :)

Reply:
Basically refactored Dictionary and used a Trie